### ISSUE TITLE
Namespaced API & enums

### ISSUE THREAD

**Issue by lstrojny** [pdezwart/php-amqp#479]
Modernize the API and introduce namespaces and enums:

For 3.x, where I would expect to only support PHP >=8.1:
 * Move all classes/interfaces into `AMQP\`
   * Provide backwards compatibility
 * Refactor some constants into sensible enums to make discoverability easier:
   * e.g. `AMQP\ExchangeType` for all  the `AMQP_EX*` constants
   * Check if we can use enums with operator overloading for flags as well (so that e.g. `AMQP\QueueFlags::IMMEDIATE | AMQP\QueueFlags::NOLOCAL` could work)
     * Should be possible, look at [gmp for inspiration](https://github.com/php/php-src/blob/master/ext/gmp/gmp.c#L373). Carefully consider if the elegant API is worth the complexity
   * Provide backwards compatibility
 * Issue deprecation warnings for old API

For 4.x:
 * Remove backwards compatibility

**Comment by lyrixx** [comment]
I would simplify this:

* 3.x
  * all what you said
  * + deprecation warning
* 4.x
  * remove BC layer

**Comment by lstrojny** [comment]
Fair enough, it probably doesn't make sense to spread out the deprecation cycle over two major versions.

**Comment by stof** [comment]
I would not try to make an enum for flags used in a bitmask: this would probably force you to define cases for all possible combination of flags so that your overloaded `|` operator can still return an instance of the enum. This does not make sense to me.

However, you might offer a QueueFlags class with int constants in it to make the API more elegant (it would still look like `AMQP\QueueFlags::IMMEDIATE | AMQP\QueueFlags::NOLOCAL` for actual usages)

**Comment by lstrojny** [comment]
> this would probably force you to define cases for all possible combination of flags so that your overloaded | operator can still return an instance of the enum. This does not make sense to me.

I haven’t look more deeply into it but my idea was to implement it a way that enum would be an int enum and that the operator overloading implementation would unwrap the value, do the bit op and wrap the value again. That being said: if the only way to get that done is to implement all combinations, then it’s not worth it.

**Comment by stof** [comment]
An enum is a list of allowed value. For a bitmask, the allowed values are all the combinations.

**Comment by lstrojny** [comment]
Yeah, you are right, the enum idea most likely won't work.

**Comment by madorin** [comment]
Any news on this?

**Comment by lstrojny** [comment]
Nothing beyond early prototypes to get the deprecation part right (PHP core does not support deprecating classes, so one needs to be a little creative with object handlers to issue deprecation warnings).
